### PR TITLE
Adding logic to add and notify subscriber

### PR DESCRIPTION
**Pull request by sethiay**
### Description
Adding helper functions for adding and notifying subscriber in download job and a function to fail while downloading. Also setup fake GCS object and bucket in job_test.go. I

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - Added unit tests for newly added functions
3. Integration tests - NA

### CODE REVIEW

File: internal/cache/file/downloader/job.go [hunk 1]
```diff
@@ -124,6 +126,47 @@ func (job *Job) Cancel() {
 	defer job.mu.Unlock()
 }
 
+// addSubscriber adds subscriber for download job and returns channel which is
+// notified when the download is completed at least till the subscribed offset
+// or in case of failure.
+//
+// Not concurrency safe and requires LOCK(job.mu)
+func (job *Job) addSubscriber(subscribedOffset int64) (notificationC <-chan JobStatus) {
```
Comment (raj-prince) [inline internal/cache/file/downloader/job.go:134]: Ideally, we should give a jobSubscriber as an input Parameter to addSubscriber. And accordingly we should not return anything. What do you think?

File: internal/cache/file/downloader/job.go [hunk 2]
```diff
@@ -124,6 +126,47 @@ func (job *Job) Cancel() {
...
+// notifySubscribers notifies all the subscribers of download job in case of
+// error/cancellation or when download is completed till the subscribed offset.
+//
+// Not concurrency safe and requires LOCK(job.mu)
+func (job *Job) notifySubscribers() {
+	subItr := job.subscribers.Front()
+	for subItr != nil {
+		subItrValue := subItr.Value.(jobSubscriber)
+		nextSubItr := subItr.Next()
+		if job.status.Name == FAILED || job.status.Name == CANCELLED || job.status.Offset >= subItrValue.subscribedOffset {
```
Comment (raj-prince) [inline internal/cache/file/downloader/job.go:149]: Just thinking out loud, the failed status here we will get after trying all the retry configured.

Comment (raj-prince) [reply internal/cache/file/downloader/job.go:149]: Can we simplify the looping over list?

Something like this,

``` go

for e := l.Front(); e != nil; e = e.Next() {
        v := e.Value
}
```

File: internal/cache/file/downloader/job.go (hunk 1 above)
Comment (sethiay) [reply internal/cache/file/downloader/job.go:134]: Would it work if we change the name from addSubscriber to subscribe and keep input and return types same ?

File: internal/cache/file/downloader/job.go [hunk 3]
```diff
@@ -124,6 +126,47 @@ func (job *Job) Cancel() {
...
+			subItrValue.notificationC <- job.status
+			close(subItrValue.notificationC)
+			job.subscribers.Remove(subItr)
+		}
+		subItr = nextSubItr
+	}
+}
+
+// errWhileDownloading changes the status of job to failed and notifies
```
Comment (raj-prince) [inline internal/cache/file/downloader/job.go:158]: nit: failWhileDownloading

File: internal/cache/file/downloader/job_test.go [hunk 4]
```diff
@@ -16,34 +16,87 @@ package downloader
...
 }
 
 func init() { RegisterTestSuite(&jobTest{}) }
 
 func (jt *jobTest) SetUp(*TestInfo) {
 	locker.EnableInvariantsCheck()
-	var fileSpec data.FileSpec
-	jt.job = NewJob(nil, nil, nil, 200, fileSpec)
+
+	jt.bucket = fake.NewFakeBucket(timeutil.RealClock(), TestBucketName)
```
Comment (raj-prince) [inline internal/cache/file/downloader/job_test.go:57]: Should we move

File: internal/cache/file/downloader/job.go (hunk 2 above)
Comment (sethiay) [reply internal/cache/file/downloader/job.go:149]: That will not work because we are removing elements while looping (if notified).

**Review by raj-prince — CHANGES_REQUESTED** [review]
Please incorporate.

File: internal/cache/file/downloader/job.go (hunk 2 above)
Comment (raj-prince) [reply internal/cache/file/downloader/job.go:149]: Still we can do this, the only thing we need to maintain a variable as we are doing above nextSubItr.

something like this,
``` go
var next *Element
for e := l.Front(); e != nil; e = next {
    next = e.Next()
    l.Remove(e)
}
```

File: internal/cache/file/downloader/job.go (hunk 1 above)
Comment (raj-prince) [reply internal/cache/file/downloader/job.go:134]: Honestly speaking, I am more inclined towards addSubscriber(jobSubscriber). In this way we give flexibility to client  to give their notificationChannel.

Comment (sethiay) [reply internal/cache/file/downloader/job.go:134]: We decided both the options are fine. Decided to go with subscribe option as the other one increase code for test.

File: internal/cache/file/downloader/job_test.go (hunk 4 above)
Comment (sethiay) [reply internal/cache/file/downloader/job_test.go]: We decided offline to keep it this way for now. (just use fake storage server bucket instead)

File: internal/cache/file/downloader/job_test.go [hunk 5]
```diff
@@ -16,34 +16,92 @@ package downloader
...
+	}
+
+	fileSpec := data.FileSpec{
+		Path:     path.Join("./", storage.TestBucketName, DefaultObjectName),
+		Perm:     os.FileMode(0666),
+		OwnerUid: uint32(os.Getuid()),
+		OwnerGid: uint32(os.Getgid()),
+	}
+
+	jt.job = NewJob(&defaultMinObject, jt.bucket, jt.cache, 200, fileSpec)
```
Comment (raj-prince) [inline internal/cache/file/downloader/job_test.go:97]: Please create a constant for sequentialReadSizeMb.

**Review by raj-prince — APPROVED** [review]

**Pull request merged**